src/quantization.cpp

```diff
@@ -568,12 +105,50 @@ capture_arguments_impl(program& prog, const target& t, const std::vector<std::st
         int8_quant_params->at(ins_index) = param_pair;
     };
 
-    auto num_params = capture_arguments(prog, ins_names, calc_quant_params);
+    // add capture argument op
+    run_passes(prog, {capture_arguments_pass{ins_names, calc_quant_params}});
 
-    int8_quant_params->resize(num_params, std::pair<float, float>(64.0f, 0.0f));
-    max_abs_vals->resize(num_params, 0.0f);
+    // use the calibration data to compute the quantization scale
+    prog.compile(t);
 
-    return int8_quant_params;
+    // use all calibration data to run the program to calculate the
+    // quantization scale and shift
+    for(auto&& arg : calibration)
+    {
+        parameter_map m;
+        for(auto&& x : prog.get_parameter_shapes())
+        {
+            if(arg.count(x.first) > 0)
+            {
+                assert(x.second == arg.at(x.first).get_shape());
+                m[x.first] = t.copy_to(arg.at(x.first));
+            }
+            else
+            {
+                m[x.first] = t.allocate(x.second);
+            }
+        }
+        prog.eval(m);
+    }
+
+    return std::move(*int8_quant_params);
+}
+
+void quantize_int8(program& prog,
+                   const target& t,
+                   const std::vector<parameter_map>& calibration,
+                   const std::vector<std::string>& ins_names)
+{
+    // insert capture operator
+    const auto& int8_quant_params = calc_quantize_params(prog, t, calibration, ins_names);
+    run_passes(prog,
+               {quantize_int8_pass{ins_names, int8_quant_params},
+                eliminate_common_subexpression{},
+                dead_code_elimination{},
+                simplify_reshapes{},
+                dead_code_elimination{},
+                simplify_qdq{},
+                dead_code_elimination{}});
```

Comment:
Why do we pass all these passes? We should run through these passes when we compile the program.

Reply:
This is to keep the quantize_int8 doing the same thing as the previous implementation. In other words, after we call this API, we get a int8 version of the program, and it can run on either gpu, cpu, or ref. With the additional passes, we are still running the program in fp32 if we not have the later passes on some target.

Reply:
But now a target doesn't have control if it wants to skip running a dot or conv as a quantized dot or conv. For example, the cpu backend doesn't run anything as int8 currently.

Reply:
We can open an issue to address this in a future PR. This doesn't really change the current behavior.

Reply:
Thanks. #943 is created to track this.